File: 2018/days/Day02.cs
namespace _2018.days;

public static class Day02
{
    public static void Run()
    {
        string[] ids = File.ReadAllLines("inputs/day02.txt");
        int twos = 0;
        int threes = 0;
        foreach (var id in ids)
        {
            var charCounts = GetCharCounts(id);
            if (charCounts.ContainsValue(2)) twos++;
            if (charCounts.ContainsValue(3)) threes++;
        }
        System.Console.WriteLine($"Part 1: {twos * threes}");

        for (int i = 0; i < ids.Length - 1; i++)
        {
            for (int j = i + 1; j < ids.Length; j++)
            {
                int diff = -1;
                bool isAnswer = true;
                string a = ids[i];
                string b = ids[j];
                for (int k = 0; k < a.Length; k++)
                {
                    if (a[k] != b[k])
                    {
                        if (diff > -1)
                        {
                            isAnswer = false;
                            break;
                        }
                        diff = k;
                    }
                }
                if (isAnswer)
                {

                    Console.WriteLine($"Part 2: {a[..diff] + a[(diff + 1)..]}");
                    goto Done;
                }
            }
        }
    Done: { }
    }

    private static Dictionary<char, int> GetCharCounts(string id)
    {
        Dictionary<char, int> charCounts = [];
        foreach (char c in id)
            charCounts[c] = charCounts.GetValueOrDefault(c, 0) + 1;
        return charCounts;
    }
}

File: 2018/day02.js
import fs from 'fs';

console.time();

const lines = fs.readFileSync("inputs/day02.txt", "utf-8").split("\n");
let twos = 0;
let threes = 0;
for (let line of lines) {
  let charCounts = getCharCounts(line);
  if (Object.values(charCounts).includes(2)) twos++;
  if (Object.values(charCounts).includes(3)) threes++;
}
console.log(`Part 1: ${twos * threes}`);
console.log(`Part 2: ${p2(lines)}`);

console.timeEnd();

function getCharCounts(str) {
  let counts = {};
  for (let i = 0; i < str.length; i++) {
    if (!counts[str[i]]) counts[str[i]] = 0;
    counts[str[i]]++;
  }
  return counts;
}

function p2(lines) {
  for (let i = 0; i < lines.length - 1; i++) {
    for (let j = i + 1; j < lines.length; j++) {
      let a = lines[i];
      let b = lines[j];
      let diffIdx = getDiffIdx(a, b);
      if (diffIdx != null) {
        return a.slice(0, diffIdx) + a.slice(diffIdx + 1);
      }
    }
  }
}

function getDiffIdx(a, b) {
  let diffIdx = null;
  for (let k = 0; k < a.length; k++) {
    if (a[k] !== b[k]) {
      if (diffIdx === null) diffIdx = k;
      else return null;
    }
  }
  return diffIdx;
}
